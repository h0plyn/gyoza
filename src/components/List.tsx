import { useState, useEffect } from 'react';
import { Asset } from '../types';
import useFetch from '../hooks/useFetch';
import Pagination from './Pagination';
import { Link } from 'react-router-dom';
import { useCoin } from '../context/coin';

export default function List() {
  const { data, loading, error } = useFetch(
    'https://api.coingecko.com/api/v3/coins/markets?vs_currency=USD&order=market_cap_desc&per_page=100&page=1&sparkline=false'
  );
  const [allCoins, setAllCoins] = useState<Asset[]>([]);
  const [page, setPage] = useState<number>(1);
  const [coinsPerPage, setCoinsPerPage] = useState<number>(10);
  const lastCoinIdx = page * coinsPerPage;
  const firstCoinIdx = lastCoinIdx - coinsPerPage;
  const { setCurrentCoin } = useCoin();

  useEffect(() => {
    if (data) setAllCoins(data);
  }, [data]);

  return (
    <div>
      {error && <div>Something went wrong fetching data...</div>}
      <table data-testid="table">
        <tbody>
          <tr>
            <th>Logo</th>
            <th>Symbol</th>
            <th>Name</th>
            <th>Current Price</th>
          </tr>
          {!loading &&
            allCoins.slice(firstCoinIdx, lastCoinIdx).map((coin: Asset) => {
              const slug = coin.id.replace(/\s+/g, '').toLowerCase();
              return (
                <tr key={coin.name} title="coins">
                  <td>
                    {' '}
                    <img
                      src={coin.image}
                      alt={coin.name}
                      style={{ maxWidth: 20, maxHeight: 20 }}
                    />
                  </td>
                  <td>${coin.symbol.toUpperCase()}</td>
                  <td onClick={() => setCurrentCoin(coin)}>
                    <Link to={`${slug}`}>{coin.name}</Link>
                  </td>
                  <td>{coin.current_price}</td>
                </tr>
              );
            })}
        </tbody>
      </table>
      <Pagination page={page} setPage={setPage} />
    </div>
  );
}
