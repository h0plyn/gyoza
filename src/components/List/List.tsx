import React, { useState, useEffect } from 'react';
import { useCoin } from '../../context/singleCoin';
import { CoinCard, Pagination, ListHeader, Search } from '../../components';
import { Asset, useSingleCoin } from '../../types';
import useFetch from '../../hooks/useFetch';
import useDebounce from '../../hooks/useDebounce';

export default function List() {
  const { data, loading, error } = useFetch(
    'https://api.coingecko.com/api/v3/coins/markets?vs_currency=USD&order=market_cap_desc&per_page=100&page=1&sparkline=false'
  );
  const [allCoins, setAllCoins] = useState<Asset[]>([]);
  const [page, setPage] = useState<number>(1);
  const [coinsPerPage, setCoinsPerPage] = useState<number>(10);
  const lastCoinIdx = page * coinsPerPage;
  const firstCoinIdx = lastCoinIdx - coinsPerPage;
  const [query, setQuery] = useState('');
  const [queryResults, setQueryResults] = useState<Asset[]>([]);
  const debouncedQuery = useDebounce(query, 500);
  const [ascending, setAscending] = useState<boolean>(true);
  const { setCurrentCoin }: useSingleCoin = useCoin();

  useEffect(() => {
    if (data && !allCoins.length) setAllCoins(data);
  }, [data, allCoins]);

  useEffect(() => {
    const searchTerm = debouncedQuery.toLowerCase();
    const results = allCoins.filter((coin: Asset) =>
      coin.name.toLowerCase().includes(searchTerm)
    );
    setQueryResults(results);
  }, [debouncedQuery, allCoins]);

  function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    setQuery(e.target.value);
  }

  function sortBy(e: React.ChangeEvent) {
    const target = e.target;
    let sorted = [...allCoins];

    if (target.innerHTML === '24hr') {
      if (ascending) {
        sorted = allCoins.sort(
          (a, b) =>
            b.price_change_percentage_24h - a.price_change_percentage_24h
        );
        setQueryResults(sorted);
        setAscending(!ascending);
      } else {
        sorted = allCoins.sort(
          (a, b) =>
            a.price_change_percentage_24h - b.price_change_percentage_24h
        );
        setQueryResults(sorted);
        setAscending(!ascending);
      }
    }
  }

  return (
    <div>
      {error && <div>Something went wrong fetching data...</div>}
      <div data-testid="container">
        <Search query={query} handleChange={handleChange} />
        <div>
          <ListHeader sortBy={sortBy} />
          {!loading &&
            queryResults
              .slice(firstCoinIdx, lastCoinIdx)
              .map((coin: Asset, idx: number) => (
                <CoinCard
                  coin={coin}
                  setCurrentCoin={setCurrentCoin}
                  idx={idx}
                  key={coin.id}
                />
              ))}
        </div>
      </div>
      <Pagination page={page} setPage={setPage} />
    </div>
  );
}
