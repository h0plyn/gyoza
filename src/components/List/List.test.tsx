import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import List from './List';

describe('<List />', () => {
  test('it should render the List component without crashing', () => {
    render(<List />)
  })

  test('it matches the snapshot of List', () => {
    const tree = renderer.create(<List />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('it displays a container to hold the List and its components', () => {
    render(<List />);
    expect(screen.getByTestId('container')).toBeInTheDocument()
  });
});
