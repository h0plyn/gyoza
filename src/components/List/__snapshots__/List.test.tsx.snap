// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> it matches the snapshot of List 1`] = `
<div>
  
  <div
    data-testid="container"
  >
    <div
      className="sc-bdnxRM edOJoy"
      data-testid="search"
    >
      <input
        data-testid="input"
        name="query"
        onChange={[Function]}
        placeholder="Coin Name"
        type="text"
        value=""
      />
    </div>
    <div>
      <div
        className="card-header"
      >
        <h3
          className="rank"
          onClick={[Function]}
        >
          Rank
        </h3>
        <h3
          className="f1"
          onClick={[Function]}
        >
          Coin
        </h3>
        <h3
          className="price"
          onClick={[Function]}
        >
          Price
        </h3>
        <h3
          className="daily-change"
          onClick={[Function]}
        >
          24hr
        </h3>
        <h3
          className="f1"
          onClick={[Function]}
        >
          Market Cap
        </h3>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "marginBottom": "3rem",
      }
    }
  >
    <div
      className="pagination"
      title="pagination"
    >
      <div
        className="pagination-numbers"
        id="current-page"
        onClick={[Function]}
      >
        1
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        2
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        3
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        4
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        5
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        6
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        7
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        8
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        9
      </div>
      <div
        className="pagination-numbers"
        id=""
        onClick={[Function]}
      >
        10
      </div>
    </div>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
